nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/text/TextLineRecordReader.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.text;
+
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.serialization.MalformedRecordException;
+import org.apache.nifi.serialization.RecordReader;
+import org.apache.nifi.serialization.SimpleRecordSchema;
+import org.apache.nifi.serialization.record.MapRecord;
+import org.apache.nifi.serialization.record.Record;
+import org.apache.nifi.serialization.record.RecordField;
+import org.apache.nifi.serialization.record.RecordFieldType;
+import org.apache.nifi.serialization.record.RecordSchema;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.List;
+import java.util.Scanner;
+
+public class TextLineRecordReader implements RecordReader {
+
+    private final ComponentLog log;
+    private final RecordSchema schema;
+    private final String fieldName;
+    private final Scanner lineReader;
+    private final String lineDelimiter;
+    private int linesLeftToSkip;
+    private final int linesPerGroup;
+    private final boolean ignoreEmptyLines;
+
+    public TextLineRecordReader(final InputStream in, final ComponentLog logger, final String fieldName,
+                                int skipLineCount, int linesPerGroup, boolean ignoreEmptyLines, String lineDelimiter, String charSet) {
+        this.log = logger;
+        this.fieldName = fieldName;
+        this.lineReader = new Scanner(in, charSet).useDelimiter("(?<=" + lineDelimiter + ")");
+        this.lineDelimiter = lineDelimiter;
+        this.linesLeftToSkip = skipLineCount;
+        // If zero, read all lines as a single record
+        this.linesPerGroup = (linesPerGroup == 0 ? Integer.MAX_VALUE : linesPerGroup);
+        this.ignoreEmptyLines = ignoreEmptyLines;
+        final List<RecordField> fieldList = Collections.singletonList(new RecordField(fieldName, RecordFieldType.STRING.getDataType()));
+        schema = new SimpleRecordSchema(fieldList);
+    }
+
```

Comment:
I know there is no requirement for java doc, but this function is pretty important, and the most likely place for maintainability.

Could you add an overview/summary of the goals, approach, logic and concerns?
It will keep folks like me messing it up